### ISSUE TITLE
support interfaces with call signatures

### ISSUE THREAD

**Issue by joshuavial** [uittorio/ts-auto-mock#81]
Great library - I've been trying to extend it to support call signatures in an interface but am getting a little lost.

This is the code that I'd like to work

```
    interface InterfaceDefiningMethod {
        (): string;
    }

    interface Interface {
        a: InterfaceDefiningMethod;
    }

    it('should set the functions', () => {
        const properties: Interface = createMock<Interface>();
        expect(properties.a()).toBe('');
    });
```

I've played around with using `typeChecker.getSignaturesOfType` to get hold of the call signature but am struggling to find the right way to return a function.

https://github.com/joshuavial/ts-auto-mock/commit/21dbe1dabb9428fe716bc133ba3b92077453b8bb has an example of what I've been trying - any tips on the best way to solve this?

**Comment by uittorio** [comment]
Hi Joshua, thank you for spending time using the library and improving it. You are right, the library doesnt currently support interface call signature, however extending it to fully support it will require a bit more work because an interface can also have properties.

Let's say we have an interface 
```ts
interface InterfaceWithCallSignature {
            (a: number): number;
            (a: string): string;
            b: string;
}
```

There are few requirements. 
1) should be callable (like you mentioned)
2) should be able to choose wich overload to use. We could support both but it wil require even more work where we will create a function that based on the arguments type will return different types
3) should have properties.

It's all possible but I would concentrate on point 1 and 2. 
Here is an example of how to support in javascript that interface 

```ts
const interfaceWithCallSignature: InterfaceWithCallSignature = Object.assign(function(a: number | string): any {
                if (typeof a === 'string') {
                    return a;
                } else {
                    return a;
                }
            }, {
                b: '',
            });

```
To reach this point we have to change the way the library generate the output. 
Right now its creating an object with getter and setter for each properties. To support this we would have to filter out the callable properties and do an object assign with the combination of possible value to return.

I will work on a POC this evening.

If you want to investigate where the exported object is generated in ts-auto-mock you can look into mockCall. thats where its receiveing the properties.

**Comment by uittorio** [comment]
@joshuavial  I've created a PR that will solve this issue. 
https://github.com/uittorio/ts-auto-mock/pull/82.

**Comment by uittorio** [comment]
I've just published a new version -> 1.2.1 that will cover this scenario. I'll leave the issue open until you've tested it :)

**Comment by joshuavial** [comment]
Yep, works perfectly. Thanks!
